doc/dev/osd_internals/snaps.rst: add clone_overlap doc (#56173)
ceph/ceph: doc/dev/osd_internals/snaps.rst

  3. Overlapping intervals between clones for tracking space usage
  4. Clone size

The *head* can't be deleted while there are still clones. Instead, it is
marked as whiteout (``object_info_t::FLAG_WHITEOUT``) in order to house the

> *SnapSet* contained in it .

Reviewer: ```suggestion
*SnapSet* contained in it.
```

snap for a particular PG.  Upon split, the prefixes to check on the parent
are adjusted such that only the objects remaining in the PG will be visible.
The children will immediately have the correct mapping.

clone_overlap
-------------

> Each SnapSet attached to the head object contains the overlapping intervals

Reviewer: The HEAD may be removed with snapshots remaining, in which case the snapset moves somewhere else.
```suggestion
Each SnapSet contains the overlapping intervals
```

Author: The head can't be removed if we still have clones, added this as well to the docs in a separate commit (a5579450853a588f17c1c405483a511d517a8abd)

Reviewer: Ah, I forgot that Sage made that change!

> between clones objects for optimising the space usage.

Reviewer 2: @zdover23  `clone objects for optimizing space`

The overlapping intervals are stored within the ``clone_overlap`` map, each element in the

> map stores the snap id and the corresponding overlap with the next newest clone.

Reviewer 2: ID, let's leave Frood out of this.

> See the following example using a 4 Byte object:

Reviewer 2: byte

object content is: [AAAA]

listnsnap output is as follows:
+------------+------------+-----------+-------------------------+
| cloneid    | snaps      | size      | overlap                 |
+============+============+===========+=========================+
| head       | -          | 4         |                         |
+------------+------------+-----------+-------------------------|

After taking a snapshot (ID 1) and re-writing the first 2 bytes of the object,

> the clone created will overlap with the new head object in its last 2 bytes.

Reviewer 2: In some places (e.g. for RGW) we write HEAD vs head, should we do so here?

Author: I changed to _head_ same as other references in this file. Moreover, the outputs from the examples attached use "head" (lower case).

> the clone created will overlap with the new head object in it's last 2 Bytes.

Reviewer 3: s/it's/its/

I will handle this in a future grammar-and-punctuation-correction commit.

Author: Ah, thanks! I'd prefer waiting for an additional approval to verify that the logic described adds up.

object     content is: [BBAA]
clone ID 1 content is: [AAAA]

+------------+------------+-----------+-------------------------+
| cloneid    | snaps      | size      | overlap                 |
+============+============+===========+=========================+
| 1          | 1          | 4         | [2~2]                   |
+------------+------------+-----------+-------------------------|
| head       | -          | 4         |                         |
+------------+------------+-----------+-------------------------|

> Following the same process, by taking another snapshot (ID: 2)

Reviewer 2: Suggest removing "Following the same process,", I think it doesn't add value.

and this time re-writing only the first 1 Byte of the object,

> the clone created (id 2) will overlap with the new head object in it's last 3 Bytes.

Reviewer 3: s/it's/its/

> While the oldest clone (id 1) will overlap with the newest clone in it's last 2 Bytes.

Reviewer 3: ibid

+------------+------------+-----------+-------------------------|
| 2          | 2          | 4         | [1~3]                   |
+------------+------------+-----------+-------------------------|
| head       | -          | 4         |                         |
+------------+------------+-----------+-------------------------|

If the head object will be completely re-written by re-writing 4 bytes,
the only existing overlap that will remain will be between the two clones.

object     content is: [DDDD]
clone ID 2 content is: [BBAA]
clone ID 1 content is: [AAAA]

+------------+------------+-----------+-------------------------+
| cloneid    | snaps      | size      | overlap                 |
+============+============+===========+=========================+
| 1          | 1          | 4         | [2~2]                   |
+------------+------------+-----------+-------------------------|
| 2          | 2          | 4         |                         |
+------------+------------+-----------+-------------------------|
| head       | -          | 4         |                         |
+------------+------------+-----------+-------------------------|

> Lastly, when removing the last snap (ID 2), no overlapping intervals will remain:

Reviewer: I think it's worth noting that that it happens not at the time the snapshot is removed, but potentially much later as part of snap trimming:
```suggestion
Lastly, after the last snap (ID 2) is removed and snaptrim kicks in, no overlapping intervals will remain:
```
